### ISSUE TITLE
Add contributing guide

### ISSUE THREAD

**Issue by ericelliott** [ericelliott/essential-javascript-links#31]
Should be based on the JSHomes Learning Resources [Contributing Guide](https://github.com/jshomes/learning-resources/blob/master/Contributing.md)

**Comment by ericelliott** [comment]
Ping @hoichi

**Comment by hoichi** [comment]
You want me to add it? I'm game.

**Comment by ericelliott** [comment]
Yeah.. you'll need to modify the text a bit, but I *do* want to use the hashtag convention and use those same hashtags, which may conflict with what we're currently doing here... that means updating our current tech specs, etc...

Does that sound ok to you?

We'll still have categories and tags. Tags are represented by one or more hashtags at the end of the entry. Categories are just titles. A single link can still belong to multiple categories.

In other words, link <-> category is many to many, link <-> hashtag is many to many. I believe that's the same as it is now, right?

**Comment by ericelliott** [comment]
See also [JSHomes Platform](Course Syllabus / lesson sequence format) todo.

**Comment by hoichi** [comment]
Well, in the JSHomes guide it looks like hashtags _are_ categories, and there's a very short list of them, so link <-> hashtag might be actually many to one.

We (per current spec) have basically three kinds of tags: categories, level and regular tags, and both categories and regular tags are many to many and can represent at least two different things: a type of content (article/video/book/tool etc.) or an domain (Node/React/IoT...).

(When I put it like that, it sounds a bit of an overkill. Anyway.)

We can probably do all three with hashtags if we reserve a few tags for level and raise a few others to a "category" status. Might be a bit user-unfriendly though: how do you know, when you add a hashtag, what is category and what isn't?

Of course, we could show _all_ the hashtags (sans level) as categories, listing some "favorites" first and then sorting the rest by the number of entries. Then any hashtag would basically mean: "also listed under:".

We could also aim at not having too much hashtags and say as much in our guides. Maybe that's the question we should start with: do we want as many tags per link as possible, or as little as possible?

> See also [JSHomes Platform](Course Syllabus / lesson sequence format) todo.

`404`. I probably have no rights to read it.

**Comment by hoichi** [comment]
What I immediately see comparing the current Essential JS Links prototype to [this](https://github.com/jshomes/learning-resources/blob/master/Contributing.md#javascript) is the lack of learning tracks that transcend category.

Which doesn't exactly help situation I'm in as a JS learner, by the by. I'm somewhat overwhelmed by the sheer amount of frameworks and tools I (probably) have to learn. Of course a great deal of thinking for myself is inevitable, but a clear path to getting up to speed and being useful would be very welcome.

Maybe there isn't _one_ path.There's backend, frontend, different tools and approaches. So maybe there should be some RPG-like skill trees.

And yes, I guess Essential JS Links is a much simpler endeavor than JSHomes, but maybe we could step back and think whether EJSL should be primarily a collection of learning tracks or an easily sortable/searchable backlog. It could be both — but that's a lot of complexity. Tricky.

On a related note, maybe a link info should also include estimated reading/watching/doing time. Also, maybe a good learning track can interleave passive learning with actual coding. It's probably not as critical for people that can apply what they learn to their own projects, but those guys probably need less guidance anyway (or at least they think they do).

**Comment by hoichi** [comment]
All right, I've looked at JSHomes somewhat more carefully. It seems that in JSHomes, categories === learning tracks, and hashtags === types of content.

For the current version of EJSL, for the moment being, a category might mean either a type of content or some domain of knowledge or a group of tools (libraries, building tools etc.) Sometimes those categories are being sorted to comprise a kind of learning track and sometimes there's no sorting order to make sense and it's just a list where you can look for things.

(We also have difficulty levels, but those can be regarded a poor man's learning tracks, in a way. Maybe there are other ways.)

Anyway, I think **the most important question** from which other decisions stem is: do we want EJSL to be:

- a collection of learning tracks (or some kind of a learning tree), like JSHomes seem to be,
- or a searchable/filterable backlog of (most of) the material a developer might need?

Could be both, I guess, but that's obviously trickier to get right. Something JSHomes-like calls for a very small set of hashtags, whereas a backlog could use as much [hash]tags as possible.

And a second, more tactical question is how those contributions are going to be submitted, technically? As GitHub comments? Or will we wait till the EJSL's own frontend is ready?

@ericelliott?

**Comment by ericelliott** [comment]
> * a collection of learning tracks (or some kind of a learning tree), like JSHomes seem to be,
> * or a searchable/filterable backlog of (most of) the material a developer might need?

Both. We can have hashtags with special meanings, such as the format (book, video, etc...), the learning levels ("Beginners", "Intermediate" and "Advanced").

> Something JSHomes-like calls for a very small set of hashtags

Nope. Just a small subset of hashtags that hold special meanings.


> how those contributions are going to be submitted, technically? As GitHub comments? Or will we wait till the EJSL's own frontend is ready?

While we're still fleshing out the specification, how they're submitted doesn't matter. Right now it's by pull request. Down the road we'll have a UI. Don't worry about the UI, yet.

**Comment by hoichi** [comment]
> While we're still fleshing out the specification, how they're submitted doesn't matter. Right now it's by pull request. Down the road we'll have a UI. Don't worry about the UI, yet.

Well, I just thought some UI features (like tags autocompletion, for instance) might affect what instructions we have to give. But you're right, of course. Details can always be added later.

**Comment by boucains** [comment]
I'm a complete noob, so please excuse me if my questions are nonsensical in this context. I'm also asking from the viewpoint of a <em>designer</em>,not a developer.

There are hundreds of small "learning-track" sites out there. Why another? This may be obvious to you, but not to me. The bigger sites have good tools, and it seems to me that you want to reinvent some part of the wheel. I'm probably wrong.

So, can someone take a sec to satisfy my curiosity? It helps me recognize things better when I see them in Github.

**Comment by ericelliott** [comment]
> There are hundreds of small "learning-track" sites out there. Why another?

The answer is in the question. There is a ton of learning material on the internet. Knowing that this one is curated by trusted experts can help users distinguish the great resources from the good resources, or even the poor resources.

**Comment by boucains** [comment]
Thanks for taking the time to answer my question.

Who is your target audience? Are the masses going to know who to trust? Are you going to be able to afford who to trust? These and many more...

I'm not asking, just thinking about what you already know - that you have a hard road ahead of you. Best of luck! Who knows, maybe I'll get a chance to take care of a minor task or two to pay back your time.

**Comment by hoichi** [comment]
Well, I've edited the Contributing Guide somewhat. Added a 'tool' hashtag, by the way. But before I make a PR, a few questions:

* How do they submit multi-word hashtags (some categories are muti-word)? With hyphens?
* Do we accept multi-type content? For instanse, an [online version](http://eloquentjavascript.net/) of Eloquent JavaScript is both a **#book** and **#exercises**. Although I don't know if there's a lot of other examples like this.
* Do we need an example of a non-learning track category?

**Comment by ericelliott** [comment]
Looks good... now we just need to move and rename the file so GitHub knows it's there. =)

It needs to be in the root directory, and called `contributing.md`.

**Comment by hoichi** [comment]
There you go: https://github.com/ericelliott/essential-javascript-links/pull/41

**Comment by jrhorn424** [comment]
Does #41 close this issue?

**Comment by ericelliott** [comment]
Yes. =)
